library(phyloseq)
library(decontam)
library(tidyverse)
library(vegan)

source("scripts/R/helpers/tpm_and_reads_per_kb.R")
source("scripts/R/helpers/decontam.R")

# THIS PATH HAS TO BE ADJUSTED BY THE USER! PANDAS AND SCIPY HAVE TO BE INSTALLED
python_path="/Users/nikolasbasler/miniforge3/envs/co_occurrence/bin/python3"

# This script uses data generated by several bash scripts:
# To get mapping stats from bam files:
#   coverage.stats.slurm
#     (calls: coverage.stats.R)
#   generate.stats.tables.slurm
# To get tax info by diamond and krona:
#   eukaryotic.diamond.slrm
#     (calls: taxID.filter.sh)
#   TaxidPath.py
#   Taxonomy.py
# To get tax info by blast and krona:
#   eukaryotic.blast.run.slrm
#   eukaryotic.blast.consolidate.slrm
#     (calls: TaxidPath.py, Taxonomy.py and taxID.filter.sh)
#   BLASTn_anicalc.py
# Also, a sample metadata file needs to be provided.

abundance.table <- read.csv("output/mapping_stats_eukvir/stats.eukvir.mapped_reads.csv.gz", row.names=1)
horizontal.cov.table <- read.csv("output/mapping_stats_eukvir/stats.eukvir.horizontal_coverage.csv.gz", row.names=1)
mean_depth_table <- read.csv("output/mapping_stats_eukvir/stats.eukvir.mean_depth.csv.gz", row.names=1)

metadata=metadata <- read.csv("data/metadata.csv")
row.names(metadata) = metadata$Sample_ID
eukaryotic_diamond <- read.csv("output/classification/ALL.SAMPLES_1kb_gnmd_cross_95-85_diamond.eukaryotic.csv", row.names = 1) %>%
  mutate(Classified_by = "diamond")
eukaryotic_blast <- read.csv("output/classification/ALL.SAMPLES_1kb_gnmd_cross_95-85_blastn.eukaryotic.csv", row.names = 1) %>%
  mutate(Classified_by = "blast")
Cressdnaviricota <- read.table("data/Cressdnaviricota.taxIDs")$V1
host_groups_df <- read.csv("data/host_groups.csv")

contig_lengths_in_kb = abundance.table %>%
  rownames() %>%
  str_split(.,"_") %>%
  sapply(., "[", 4) %>% 
  str_replace(".*[HV]","") %>%
  as.numeric()/1000
contig_length_df = data.frame(contig=rownames(abundance.table), length_kb=contig_lengths_in_kb)

# Min contig length filter
min_kb_filter = 3
abundance.table <- contig_length_df %>%
  filter(length_kb >= min_kb_filter) %>%
  inner_join(., rownames_to_column(abundance.table, "contig"), by="contig") %>% 
  select(-length_kb) %>%
  column_to_rownames("contig")


# Filter blast results by query coverage
qcov_filter = 50
eukaryotic_blast_qfilt = eukaryotic_blast %>%
  filter(qcov>=qcov_filter)

# Combine remaining blast results with diamond results
unique_to_blast = setdiff(row.names(eukaryotic_blast_qfilt), row.names(eukaryotic_diamond))
euk_diamond_and_blast = eukaryotic_blast_qfilt[unique_to_blast,] %>%
  select(-tname, -num_alns, -pid, -qcov, -tcov) %>%
  rbind(eukaryotic_diamond, .)

# Filter out some phages that are still in the list for some reason
filter_out_families <- c("Ackermannviridae", "Autographiviridae", 
                         "Drexlerviridae", "Herelleviridae", 
                         "Mesyanzhinovviridae", "Microviridae", 
                         "Rountreeviridae", "Straboviridae", "Zobellviridae",
                         "Autolykiviridae") # Autolykiviridae was classified by diamond. They are vibrio phages that are in division 9 for some reason...
filter_out_orders <- c("Crassvirales")
filter_out_classes <- c("Caudoviricetes", "Leviviricetes")

# euk_dia_bla_refined %>%
#   filter(if_all(c(Phylum, Class, Order, Family, Subfamily, Genus), ~. =="")) %>%
#   rownames_to_column("contig") %>% 
#   select(taxID, Species) %>% 
#   distinct() %>%
#   filter(str_detect(Species, fixed("sp.")))

# These are manually reviewed from the species list of all classifications
# It concerns most of the species classifications that have
# a "sp." in their name.
filter_out_species <- c("Riboviria sp.", "Virus sp.", 
                        "Circular genetic element sp.", "ssDNA virus sp.",
                        "Circular ssDNA virus sp.", 
                        "Prokaryotic dsDNA virus sp.") # taxID 2591644. No idea why this is here. NCBI taxonomy considers this to be part of division 9 "Viruses", instead of 3 "Phages" or 11 "Environmental samples"

euk_dia_bla_refined <- euk_diamond_and_blast %>% 
  filter(!Species %in% filter_out_species,
         !Family %in% filter_out_families,
         !Order %in% filter_out_orders,
         !Class %in% filter_out_classes) %>%
  filter(!if_all(c(Phylum, Class, Order, Family, Subfamily, Genus, Species), ~. ==""))

##################################################################
# Filter down the abundance, hzc amd mean depth tables
abundance_table_euk <- rownames_to_column(abundance.table, "contig") %>%
  filter(contig %in% row.names(euk_dia_bla_refined)) %>%
  column_to_rownames("contig")
hzc_table_euk <- rownames_to_column(horizontal.cov.table, "contig") %>%
  filter(contig %in% row.names(abundance_table_euk)) %>%
  column_to_rownames("contig")
mean_depth_table_euk <- rownames_to_column(mean_depth_table, "contig") %>%
  filter(contig %in% row.names(abundance_table_euk)) %>%
  column_to_rownames("contig")

cat(paste0(nrow(euk_dia_bla_refined),
          ' eukaryotic contigs with a classification on Phylum, Class, Order, Family, Genus or Species level.',
          'of which ', nrow(abundance_table_euk), ' are above the threshold of ', min_kb_filter, 'kb.'))
rownames_to_column(euk_dia_bla_refined, "contig") %>%
  inner_join(., rownames_to_column(abundance_table_euk, "contig"), by="contig") %>% 
  select(Classified_by) %>%
  table()


# Filtering by horizontal coverage and mean depth
hzc.filter=50
mean.depth.filter = 1

if (!identical(rownames(abundance_table_euk), rownames(hzc_table_euk)) |
    !identical(colnames(abundance_table_euk), colnames(hzc_table_euk)) |
    !identical(rownames(abundance_table_euk), rownames(mean_depth_table_euk)) |
    !identical(colnames(abundance_table_euk), colnames(mean_depth_table_euk))) 
{
  stop("Row or column names in hzc or mean depth table don't match up!")
}

abundance_table_euk[hzc_table_euk < hzc.filter]=0
abundance_table_euk[mean_depth_table_euk < mean.depth.filter]=0

# Filter out contaminants
decontam <- decontam_identification(abtable=abundance_table_euk, 
                                    classification=euk_dia_bla_refined)
# The only found contaminant contig (3kb lentgh filter) is a Lake sinai virus, 
# so it's not filtered out!
# abundance_table_euk <- abundance_table_euk %>% 
#   rownames_to_column("contig") %>%
#   filter(!contig %in% decontam$contaminants) %>%
#   column_to_rownames("contig")


# Filter out contigs that, after the hzc filter, are only present in blanks,
# then filter out contigs and samples that only have zeros.
contigs_before_hzc = rownames(abundance_table_euk)
samples_before_hzc = colnames(abundance_table_euk)

contigs_only_in_blanks <- abundance_table_euk %>%
  select(!contains("Blank")) %>%
  filter(rowSums(.)==0) %>%
  rownames()
abundance_table_euk <- abundance_table_euk %>%
  rownames_to_column("contig") %>%
  filter(!contig %in% contigs_only_in_blanks) %>%
  column_to_rownames("contig") %>%
  select(!names(.)[colSums(.) == 0]) %>%
  filter(rowSums(.)>0) 

contigs_after_hzc = rownames(abundance_table_euk)
removed_contigs = setdiff(contigs_before_hzc, contigs_after_hzc)

samples_after_hzc = colnames(abundance_table_euk)
removed_samples = setdiff(samples_before_hzc, samples_after_hzc)

cat(length(contigs_before_hzc)-length(contigs_after_hzc), "of", length(contigs_before_hzc),
          "contigs and", length(samples_before_hzc)-length(samples_after_hzc), "of", length(samples_before_hzc),
          "samples removed due to horizontal coverage filter of", hzc.filter,", mean depth filter of", mean.depth.filter, "or because only present in blanks.\n",
          "\nRemoved contigs:\n", removed_contigs, "\n",
          "\nRemoved samples:\n", removed_samples)

cat("Lowest number of mapped reads (>0) to any contig:", min(abundance_table_euk[abundance_table_euk>0]),"\n")

################################################################################
################################################################################
## Co-occurrence analysis

system("mkdir -p output/R/co_occurrence/")
abundance_table_euk %>%
  rownames_to_column("contig") %>% 
  select(contig, contains("_rec_")) %>% 
  arrange(contig) %>%
  write_tsv("output/R/co_occurrence/euk.abundances.for.co-occ.tsv")

contig_length_df %>%
  filter(contig %in% rownames(abundance_table_euk)) %>%
  arrange(contig) %>%
  write_tsv("output/R/co_occurrence/euk.contig.lengths.tsv", col_names = FALSE)

system(paste0(python_path,
              " scripts/R/helpers/co-occurrence.py",
              " --input output/R/co_occurrence/euk.abundances.for.co-occ.tsv",
              " --lengths output/R/co_occurrence/euk.contig.lengths.tsv",
              " --output output/R/co_occurrence/co_occurrences",
              " --correlation 0.5"))
co_occurrences_related_contigs <- read.delim("output/R/co_occurrence/co_occurrences_related_contigs.tsv")

co_occ_join <- co_occurrences_related_contigs %>%
  inner_join(., rownames_to_column(euk_dia_bla_refined, "contig"), by=join_by(Contig1==contig)) %>%
  inner_join(., rownames_to_column(euk_dia_bla_refined, "contig"), by=join_by(Contig2==contig)) %>%
  select(-contains(c("Avg_log_e.value", "Classification", 
                     "Classified_by"))) %>%
  rename(taxID_Contig1 = taxID.x, taxID_Contig2 = taxID.y,
         Phylum_Contig1 = Phylum.x, Phylum_Contig2 = Phylum.y,
         Class_Contig1 = Class.x, Class_Contig2 = Class.y,
         Order_Contig1 = Order.x, Order_Contig2 = Order.y,
         Family_Contig1 = Family.x, Family_Contig2 = Family.y,
         Subfamily_Contig1 = Subfamily.x, Subfamily_Contig2 = Subfamily.y,
         Genus_Contig1 = Genus.x, Genus_Contig2 = Genus.y,
         Species_Contig1 = Species.x, Species_Contig2 = Species.y)
  
co_occ_stats <- co_occ_join %>%
  select(Family_Contig1, Species_Contig1, Family_Contig2, Species_Contig2) %>%
  group_by_all() %>%
  mutate(number_of_correlations=n()) %>% 
  unique() %>%
  arrange(desc(number_of_correlations))

################################################################################
################################################################################
# TPM and prevalence filter

euk_contig_tpm_temp <- rownames_to_column(abundance_table_euk, "contig") %>%
  inner_join(., contig_length_df, by="contig") %>%
  calc_tpm(., "contig")
  
tpm_thresh <- 0.05
prev_thresh <- 1

passed_tpm_and_prevalence_filters <- euk_contig_tpm_temp %>%
  select(!contains("Blank")) %>%
  column_to_rownames("contig") %>%
  filter(if_any(everything(), ~ . >= tpm_thresh)) %>%
  filter(rowSums(. > 0) >= prev_thresh) %>%
  rownames_to_column("contig") %>%
  inner_join(., rownames_to_column(euk_dia_bla_refined, "contig"), by="contig") %>%
  select(contig) %>%
  unlist(use.names = FALSE)

euk_dia_bla_refined <- rownames_to_column(euk_dia_bla_refined, "contig") %>%
  filter(contig %in% passed_tpm_and_prevalence_filters) %>%
  column_to_rownames("contig")

abundance_table_euk <- rownames_to_column(abundance_table_euk, "contig") %>%
  filter(contig %in% passed_tpm_and_prevalence_filters) %>%
  column_to_rownames("contig")

##
max_tpm_and_prev_df <- euk_contig_tpm_temp %>%
  select(!contains("Blank")) %>%
  rowwise() %>%
  mutate(max_tpm = max(c_across(-contig)),
         mean_tpm_where_present = rowMeans(across(-c(contig, max_tpm), ~ifelse(. > 0, ., NA)), na.rm = TRUE),
         prevalence = sum(c_across(-c(contig, max_tpm, mean_tpm_where_present)) > 0),
         prevalence_of_tpm_5perc = sum(c_across(-c(contig, max_tpm, mean_tpm_where_present, prevalence)) > 0.05)) %>%
  select(contig, max_tpm, mean_tpm_where_present, prevalence, prevalence_of_tpm_5perc) %>%
  inner_join(rownames_to_column(euk_dia_bla_refined, "contig"),. , by="contig")

max_tpm_and_prev_hist <- max_tpm_and_prev_df %>%
  select(contig, max_tpm, mean_tpm_where_present, prevalence, prevalence_of_tpm_5perc) %>%
  pivot_longer(-contig) %>%
  ggplot(aes(x=value)) +
  geom_histogram(bins=50) +
  facet_wrap(~name, scales="free")


################################################################################
################################################################################
# Collapse AMFV contigs and append the artificial entry to dataframes

AMFV_contigs <- rownames_to_column(euk_dia_bla_refined, "contig") %>%
  filter(Species=="Apis mellifera filamentous virus") %>%
  select(contig) %>%
  unlist(use.names = FALSE)

AMFV_total_length <- contig_length_df %>%
  filter(contig %in% AMFV_contigs) %>%
  select(length_kb) %>%
  summarise(sum(length_kb)) %>%
  unlist(use.names = FALSE)

AMFV_supercontig_name <- paste0("AMFV_allcontigs_length_", AMFV_total_length*1000, "_cov_NA_xx_xxxxx_xxx_xxx_x")
contig_length_df <- add_row(contig_length_df, contig = AMFV_supercontig_name, length_kb=AMFV_total_length)

euk_dia_bla_refined <- rownames_to_column(euk_dia_bla_refined, "contig") %>%
  add_row(contig = AMFV_supercontig_name, taxID = 1100043,
          Classification = "Viruses", Phylum = "", Class = "", Order = "",
          Family = "", Subfamily = "", Genus = "",
          Species = "Apis mellifera filamentous virus (all contigs)",
          Classified_by = "diamond") %>%
  column_to_rownames("contig")

abundance_table_euk <- rownames_to_column(abundance_table_euk, "contig") %>%
  filter(contig %in% AMFV_contigs) %>%
  pivot_longer(-contig) %>%
  group_by(name) %>%
  mutate(value = sum(value), contig=AMFV_supercontig_name) %>%
  unique() %>%
  pivot_wider() %>%
  bind_rows(rownames_to_column(abundance_table_euk, "contig"), .) %>%
  filter(!contig %in% AMFV_contigs) %>%
  column_to_rownames("contig")

################################################################################
################################################################################
## Add host group information to classification table

euk_dia_bla_refined <- host_groups_df %>%
  select(Rank, Taxon, Host_group) %>%
  left_join(rownames_to_column(euk_dia_bla_refined, "contig"), ., by=join_by(Family==Taxon)) %>%
  
  left_join(., host_groups_df, by=join_by(Species==Taxon)) %>% 
  mutate(Host_group = coalesce(Host_group.x, Host_group.y)) %>% 
  select(-c(Host_group.x, Host_group.y),
         -contains(c("Rank", "Associated_family","Comment"))) %>% 
  
  mutate(Host_group = ifelse(taxID %in% Cressdnaviricota, 
                             "unknown_or_nonsensical", Host_group)) %>%
  column_to_rownames("contig") %>%
  mutate(Host_group = ifelse(is.na(Host_group), "unassigned", Host_group))

################################################################################
################################################################################
# Generate absolute count table .Be aware that most eukaryotic viruses will have
# an RNA genome and virus counting was done with a flow cytometer and SYBR green
# staining. RNA genomes will not stain well with SYBER green.

samples_with_vlp_counts = metadata %>%
  filter(!is.na(VLPs_per_ul)) %>%
  rownames()

viral_loads <- euk_contig_tpm_temp %>%
  select(contig, all_of(samples_with_vlp_counts)) %>%
  filter(!if_all(-contig, ~. == 0)) %>%
  pivot_longer(-contig) %>%
  rename(Sample_ID = name) %>%
  left_join(., metadata[c("Sample_ID", "VLPs_per_ul")], by="Sample_ID") %>%
  mutate(viral_load = value * VLPs_per_ul) %>%
  select(-c(value, VLPs_per_ul)) %>%
  pivot_wider(names_from = Sample_ID, values_from = viral_load)

###

subfam_but_no_fam = euk_dia_bla_refined %>%
  select(Family, Subfamily) %>%
  filter(Subfamily!="unclassified" & Family=="unclassified") %>%
  nrow()

taxlevels <- c("Species", "Genus", "Subfamily", "Family", "Order", "Class", "Phylum")
for (lvl in taxlevels) {
  number <- euk_dia_bla_refined %>%
    select(all_of(lvl)) %>%
    filter(.data[[lvl]]!="") %>%
    unique() %>% nrow()
    
  cat(paste0(number," ", lvl, "-level classifications.\n"))
}
cat(paste("There are", subfam_but_no_fam, "subfamily designations without a corresponding family."))


################################################################################
################################################################################

# Write output files

# Decontam graph
ggsave("output/R/decontam.library.size.by.control.or.sample.pdf", decontam$plot, width=30, height=10)

# Viral load table
system("mkdir -p output/R/euk_filtered/")
write_csv(viral_loads, file="output/R/euk_filtered/euk.viral.load.VLPs.per.ul.csv")

# Eukaryotic viral classification table
rownames_to_column(euk_dia_bla_refined, "contig") %>%
  inner_join(., contig_length_df, by="contig") %>% 
  write_csv(., 
          file="output/R/euk_filtered/euk.diamond.and.blast.classification.csv")

# Co-occurrence
write.csv(co_occ_join, "output/R/co_occurrence/co_occ_joined.csv", row.names=FALSE)
write.csv(co_occ_stats, "output/R/co_occurrence/co_occ_stats.csv", row.names=FALSE)

# Max TPM and prevalance
write.csv(max_tpm_and_prev_df, "output/R/max_tpm_and_prev.csv", row.names=FALSE)
ggsave("output/R/max_tpm_and_prev.pdf", max_tpm_and_prev_hist, width=8, height=5)


# Raw abundance tables
write_csv(rownames_to_column(abundance_table_euk, "contig"), file = "output/R/euk_filtered/euk.abundance.contig.csv")

